### ISSUE TITLE
`cannot replace slice of data that was already replaced` when trying to apply lifetime elision suggestions

### ISSUE THREAD

**Issue by FrancisRussell** [rust-lang/rust-clippy#16316]
### Summary

When attempting to apply Clippy suggestions regarding lifetime elision, the following error is generated:

```
warning: error applying suggestions to `src/header_rewriter.rs`

The full error message was:

> cannot replace slice of data that was already replaced

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag
```

Additionally, it's not obvious to me that the lifetimes can actually be elided though it's possible Rust has got cleverer.

### Reproducer

```bash
$ git clone https://github.com/FrancisRussell/zoog.git
$ cd zoog
$ git checkout 6c202873f9433ce4e96376857d3fd4a0038fe70e
$ cargo clippy --fix
```

### Version

```text
rustc 1.94.0-nightly (56f24e00c 2025-12-29)
binary: rustc
commit-hash: 56f24e00c62ec2db9a8b407e7aef532caaf71062
commit-date: 2025-12-29
host: x86_64-unknown-linux-gnu
release: 1.94.0-nightly
LLVM version: 21.1.8
```

**Comment by ada4a** [comment]
Minimal reproducer:

```rs
#![warn(clippy::elidable_lifetime_names)]

use derivative::Derivative;

#[derive(Derivative)]
#[derivative(Debug)]
struct HeaderRewriter<'a>(&'a ());

fn main() {}
```

The error seems to come from the `Debug` impl created by `derivative` (using `#[derivative(Debug)]`).

To fix this, the `elidable_lifetime_names` should avoid linting code coming from proc-macros (which is almost definitively what `derivative` is using). For further info, see https://doc.rust-lang.org/clippy/development/macro_expansions.html#the-is_from_proc_macro-function. Importantly, the `is_from_proc_macro` check should ideally be the last check in the lint, as it's quite expensive.

@rustbot label T-macros "good first issue"

**Comment by ada4a** [comment]
@FrancisRussell for some reason, the log snippet you provided wasn't complete -- the full version looks like this:

<details><summary>Details</summary>
<p>

```
warning: error applying suggestions to `src/comment_rewrite.rs`

The full error message was:

> cannot replace slice of data that was already replaced

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

warning: error applying suggestions to `src/header_rewriter.rs`

The full error message was:

> cannot replace slice of data that was already replaced

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

       Fixed src/comment_rewrite.rs (0 fixes)
       Fixed src/header_rewriter.rs (0 fixes)
warning: the following explicit lifetimes could be elided: 'a
   --> src/header_rewriter.rs:156:27
    |
156 | pub struct HeaderRewriter<'a, HR: HeaderRewrite, HS: HeaderSummarize, W: Write, E> {
    |                           ^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.92.0/index.html#elidable_lifetime_names
note: the lint level is defined here
   --> src/lib.rs:1:9
    |
  1 | #![warn(clippy::pedantic)]
    |         ^^^^^^^^^^^^^^^^
    = note: `#[warn(clippy::elidable_lifetime_names)]` implied by `#[warn(clippy::pedantic)]`
help: elide the lifetimes
    |
156 - pub struct HeaderRewriter<'a, HR: HeaderRewrite, HS: HeaderSummarize, W: Write, E> {
156 + pub struct HeaderRewriter<'_, HR: HeaderRewrite, HS: HeaderSummarize, W: Write, E> {
    |

warning: the following explicit lifetimes could be elided: 'a
  --> src/comment_rewrite.rs:10:32
   |
10 | pub enum CommentRewriterAction<'a> {
   |                                ^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.92.0/index.html#elidable_lifetime_names
help: elide the lifetimes
   |
10 - pub enum CommentRewriterAction<'a> {
10 + pub enum CommentRewriterAction'_> {
   |

warning: `zoog` (lib) generated 2 warnings (run `cargo clippy --fix --lib -p zoog` to apply 2 suggestions)
       Fixed src/bin/opusgain.rs (1 fix)
warning: this function has too many lines (127/100)
   --> src/bin/zoogcomment.rs:269:1
    |
269 | fn main_impl() -> Result<(), AppError> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.92.0/index.html#too_many_lines
note: the lint level is defined here
   --> src/bin/zoogcomment.rs:1:9
    |
  1 | #![warn(clippy::pedantic)]
    |         ^^^^^^^^^^^^^^^^
    = note: `#[warn(clippy::too_many_lines)]` implied by `#[warn(clippy::pedantic)]`

warning: `zoog` (bin "zoogcomment") generated 1 warning
```

</p>
</details>

Crucially, it contains the spans for the problematic lint. Having this would've made it easier to know where to look, and would've saved me from needing to diagnose the errors in the (admittedly pretty well-documented) build script of `audiopus-sys`...
